### PR TITLE
feat(rest): Added creating user from REST and basic authentication

### PR DESCRIPTION
**Pull request by smrutis1**
Signed-off-by: Smruti Prakash Sahoo <smruti.sahoo@siemens.com>

[//]: # (Copyright Bosch.IO GmbH 2020)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

> Please provide a summary of your changes here.

* Added a password field in the user structure.
* Added username and password based validation from REST endpoint.
* Added one endpoint for creating user. 
* 
Note: The above endpoints are required for building new UI and initial step to get rid of Liferay.

> * Which issue is this pull request belonging to and how is it solving it? (*#1826*)
> * Did you add or update any new dependencies that are required for your change? - No

Issue: 

* Need to test creating `/api/users` endpoint for creating user.
* And validating the authentication for that user from REST endpoint.

> Have you implemented any additional tests?

### Checklist
Must:
- [ ] All related issues are referenced in commit messages and in PR

### CODE REVIEW

**Comment by rudra-superrr** [comment]
Three things that I observed during testing are:

1. ExternalId/GID value is not being stored in the database and make this value as mandatory.
2. Through this rest endpoint password field is getting added in the sw360users database which was not there before.
3. Make all the fields used for user creation in body section as mandatory which is not the case right now.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/user/UserController.java [hunk 1]
```diff
@@ -12,6 +12,8 @@
 import lombok.NonNull;
 import lombok.RequiredArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
+
+import org.eclipse.sw360.datahandler.thrift.AddDocumentRequestSummary;
```
Comment (akapti) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/user/UserController.java:16]: remove unused import

**Review by akapti — CHANGES_REQUESTED** [review]
kindly incorporate review comments.

File: backend/src/src-users/src/main/java/org/eclipse/sw360/users/db/UserDatabaseHandler.java [hunk 2]
```diff
@@ -76,12 +78,25 @@ private void prepareUser(User user) throws SW360Exception {
         ThriftValidate.prepareUser(user);
     }
 
-    public RequestStatus addUser(User user) throws SW360Exception {
+    public AddDocumentRequestSummary addUser(User user) throws SW360Exception {
         prepareUser(user);
+        AddDocumentRequestSummary addDocReqSummarry = new AddDocumentRequestSummary();
+        if(isDuplicate(user)) {
+            return addDocReqSummarry.setRequestStatus(AddDocumentRequestStatus.DUPLICATE);
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply backend/src/src-users/src/main/java/org/eclipse/sw360/users/db/UserDatabaseHandler.java]: Added user id in the response

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/security/basic/Sw360CustomUserDetailsService.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        Authentication clientAuthentication = SecurityContextHolder.getContext().getAuthentication();
+        org.springframework.security.core.userdetails.User result = null;
+        User user = sw360UserDetailsProvider.provideUserDetails(userid, null);
+        result = new org.springframework.security.core.userdetails.User(user.getEmail(), user.getPassword(),
+                Sw360GrantedAuthoritiesCalculator.generateFromUser(user));
+
+        if (result != null) {
+            return result;
+        } else {
+            throw new UsernameNotFoundException("No user with username " + user.getEmail() + " found in sw360 users.");
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/security/basic/Sw360CustomUserDetailsService.java]: removed

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/Sw360ResourceServer.java [hunk 4]
```diff
@@ -66,6 +71,8 @@ public class Sw360ResourceServer extends SpringBootServletInitializer {
         JWKS_ISSUER_URL = props.getProperty("jwks.issuer.url", null);
         JWKS_ENDPOINT_URL = props.getProperty("jwks.endpoint.url", null);
         IS_JWKS_VALIDATION_ENABLED = Boolean.parseBoolean(props.getProperty("jwks.validation.enabled", "false"));
+        CONFIG_WRITE_ACCESS_USERGROUP = UserGroup.valueOf(props.getProperty("rest.write.access.usergroup", DEFAULT_WRITE_ACCESS_USERGROUP));
+        CONFIG_ADMIN_ACCESS_USERGROUP = UserGroup.valueOf(props.getProperty("rest.admin.access.usergroup", DEFAULT_ADMIN_ACCESS_USERGROUP));
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/Sw360ResourceServer.java:75]: Added

**Comment by smrutis1** [comment]
@rudra-superrr & @akapti Thank you for review and test. I have incorporated the changes. Kindly have a look once.

**Comment by rudra-superrr** [comment]
Things that I observed during testing are:

1. Make userGroup and externalId field mandatory in request structure/body.
2. Getting 500 error code for some fields 
image

and 400 error code for some fields
image

Error code(400)  should be same if any of the field is not added in body.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/security/basic/Sw360CustomUserDetailsService.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.eclipse.sw360.datahandler.thrift.users.User;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Profile;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
```
Comment (akapti) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/security/basic/Sw360CustomUserDetailsService.java:16]: remove unused imports. line `12` and `13` as well.

Comment (smrutis1) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/security/basic/Sw360CustomUserDetailsService.java]: Removed

File: backend/src/src-users/src/main/java/org/eclipse/sw360/users/db/UserDatabaseHandler.java [hunk 6]
```diff
@@ -76,12 +82,27 @@ private void prepareUser(User user) throws SW360Exception {
...
     }
 
-    public RequestStatus addUser(User user) throws SW360Exception {
+    public AddDocumentRequestSummary addUser(User user) throws SW360Exception {
         prepareUser(user);
+        AddDocumentRequestSummary addDocReqSummarry = new AddDocumentRequestSummary();
+        User existingUserInDB = getByEmail(user.getEmail());
+
+        if (CommonUtils.isNullEmptyOrWhitespace(user.getEmail())) {
+            return addDocReqSummarry.setMessage(EMAIL_ID_IS_MANDATORY).setRequestStatus(AddDocumentRequestStatus.INVALID_INPUT);
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply backend/src/src-users/src/main/java/org/eclipse/sw360/users/db/UserDatabaseHandler.java]: (from discussion):
`userGroup` should be by default `USER` and `externaiid` is not mandatory.

File: backend/src/src-users/src/main/java/org/eclipse/sw360/users/UserHandler.java [hunk 7]
```diff
@@ -101,9 +102,8 @@ public List<User> getAllUsers() {
     }
 
     @Override
-    public RequestStatus addUser(User user) throws TException {
+    public AddDocumentRequestSummary addUser(User user) throws TException {
         assertNotNull(user);
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply backend/src/src-users/src/main/java/org/eclipse/sw360/users/UserHandler.java]: Used.

**Comment by smrutis1** [comment]
@rudra-superrr & @akapti Please have a look again

**Review by akapti — APPROVED** [review]
Code looks good to me.
@rudra-superrr  Can you please re-test it once?

**Review by heliocastro — APPROVED** [review]
Good to go

**Comment by rudra-superrr** [comment]
Hi @smrutis1,
- Can you remove userGroup parameter from doc as this field is now not mandatory and user will be provided with user role by default.
- Other thing, just change 500 to 400 error code for password field too.
image

**Comment by smrutis1** [comment]
> Hi @smrutis1,
> 
> * Can you remove userGroup parameter from doc as this field is now not mandatory and user will be provided with user role by default.
> * Other thing, just change 500 to 400 error code for password field too.
>   image

Done

**Comment by rudra-superrr** [comment]
Testing was successful.

**Comment by ag4ums** [comment]
@smrutis1, need to separate the PR from current implementation with a switch, else anyone in production servers can use this endpoint to login to backend

**Comment by smrutis1** [comment]
> @smrutis1, need to separate the PR from current implementation with a switch, else anyone in production servers can use this endpoint to login to backend

@ag4ums Added the flag in `application.yml` to disable/enable the endpoint. The user will get a `Service is disabled message` when it is disabled.

image

**Comment by rudra-superrr** [comment]
Testing was successful.
Getting the below message when service is disabled i.e. set to true.

image

**Review by ag4ums — APPROVED** [review]
Changes looks good

**Pull request merged**